src/layers/basic.jl

```diff
@@ -503,3 +503,61 @@ function Base.show(io::IO, e::Embedding)
 end
 
 outputsize(e::Embedding) = (e.out_dims,)
+
+"""
+    PeriodicEmbedding(dims, periods)
+
+Create an embedding periodic in some dimensions with specified periods. Dimensions not in
+`dims` are passed through unchanged, but dimensions in `dims` are moved to the end of the
+output and replaced with their sines, followed by their cosines (scaled appropriately to
+have the specified periods). This smooth embedding preserves phase information and enforces
+periodicity.
+
+## Arguments
+
+  - `dims`: Periodic dimensions
+  - `periods`: Periods of the dimensions `dims`, in the same order
+
+## Inputs
+
+  - `x` must be an `AbstractArray` with `issubset(dims, axes(x, 1))`
+  - `st` must be a `NamedTuple` where `st[:k] = 2 ./ periods`, but is on the same device as
+    `x`
+
+## Returns
+
+  - `AbstractArray` of size `(size(x, 1) + length(dims), ...)` where `...` are the other
+    dimensions of `x`.
+  - `NamedTuple` with field `k` equal to `st[:k]`, but with elements of the same type as the
+    elements of `x`
+"""
+@concrete struct PeriodicEmbedding <:AbstractExplicitLayer
+    dims
+    periods
+end
+
+Lux.initialstates(::AbstractRNG, p::PeriodicEmbedding) = (k = 2 ./ p.periods,)
+
+@inline function (p::PeriodicEmbedding)(x::AbstractVector, ps, st::NamedTuple)
+    return vec(first(p(reshape(x, :, 1), ps, st))), st
+end
+
+@inline function (p::PeriodicEmbedding)(x::AbstractMatrix{T}, ps, st::NamedTuple) where T
+    other_dims = ChainRulesCore.@ignore_derivatives setdiff(axes(x, 1), p.dims)
+    _st = (k = T.(st[:k]), )
+    return (
+        vcat(
+            view(x, other_dims, :),
+            sinpi.(_st[:k].* view(x, p.dims, :)),
+            cospi.(_st[:k] .* view(x, p.dims, :))
+        ),
+        _st)
```

Comment:
I would not return the type-casted `st`. changing type of `st` is permitted but makes life much harder when used with StatefulLuxLayer